site/app/controllers/admin/ReportController.php

```diff
@@ -654,62 +692,184 @@ public function generateCustomization() {
-                JsonResponse::getErrorResponse($msg),
-                null,
-                new RedirectResponse($redirect_url)
-            );
+            return JsonResponse::getErrorResponse($msg);
         }
 
         $upload = $_FILES['config_upload'];
-        if (empty($upload['tmp_name'])) {
+        if (!isset($upload['tmp_name']) || trim($upload['tmp_name']) === '') {
             $msg = 'Upload failed: Empty tmp name for file';
             $this->core->addErrorMessage($msg);
+            return JsonResponse::getErrorResponse($msg);
+        }
+
+        $rainbow_grades_dir = FileUtils::joinPaths($this->core->getConfig()->getCoursePath(), "rainbow_grades");
+        $destination_path = FileUtils::joinPaths($rainbow_grades_dir, 'manual_customization.json');
+
+        // this is changed from move_uploaded_file to copy because of permission issue (bits not carried over)
+        // copy is expensive, but we are OK because it is small file.
+        // setgid (sticky-bit) gets ignored and doesn't inherit the parent (rainbowgrades dir) permissions
+        // known issue: look https://www.php.net/manual/en/function.move-uploaded-file.php for more details
+        if (!copy($upload['tmp_name'], $destination_path)) {
+            $msg = 'Upload failed: Could not copy file';
+            $this->core->addErrorMessage($msg);
+            return JsonResponse::getErrorResponse($msg);
+        }
+
+        $manual_customization_exists =  file_exists($destination_path);
+
+        $msg = 'Rainbow Grades Customization uploaded';
+        $this->core->addSuccessMessage($msg);
+
+        return JsonResponse::getSuccessResponse([
+            'customization_path' => $rainbow_grades_dir,
+            'manual_customization_exists' => $manual_customization_exists
+        ]);
+    }
+
+
+    #[Route("/courses/{_semester}/{_course}/reports/rainbow_grades_customization/manual_download", methods: ["GET"])]
+    public function downloadRainbowConfig() {
+        $rainbow_grades_dir = FileUtils::joinPaths($this->core->getConfig()->getCoursePath(), "rainbow_grades");
+        $file_path = FileUtils::joinPaths($rainbow_grades_dir, 'manual_customization.json');
+
+        if (!file_exists($file_path)) {
+            $msg = 'Download failed: File not found';
+            $this->core->addErrorMessage($msg);
+            $redirect_url = $this->core->buildCourseUrl(['reports', 'rainbow_grades_customization']);
             return new MultiResponse(
                 JsonResponse::getErrorResponse($msg),
                 null,
                 new RedirectResponse($redirect_url)
             );
         }
 
+        // Set headers and read the file
+        header('Content-Type: application/json');
+        header("Content-Transfer-Encoding: Binary");
+        header("Content-disposition: attachment; filename=\"" . basename($file_path) . "\"");
+
+        readfile($file_path);
+        exit;
+        // Do not add return type, for example if you add :MultiResponse
```

Comment:
Could this use the DownloadResponse instead of setting headers manually? (Same with the next download)